File: packages/sdk-runtime/src/privacy.ts
import type { JsonObject, JsonValue } from './types';

export const TELEGRAM_CONTACT_FIELDS = [
  'username',
  'firstName',
  'lastName',
  'languageCode',
  'isPremium',
  'addedToAttachmentMenu',
] as const;

export type TelegramContactField = (typeof TELEGRAM_CONTACT_FIELDS)[number];
export type MessageContentMode = 'full' | 'routing' | 'none';
export type TelegramTextField =
  | 'messageText'
  | 'caption'
  | 'commandArguments'
  | 'callbackData'
  | 'inlineQuery'
  | 'contactVCard'
  | 'buttonText';

export interface TelegramTextTransformContext {
  value: string;
  field: TelegramTextField;
  updateType: string;
}

export type TelegramTextTransformer = (
  context: TelegramTextTransformContext,
) => string | null | Promise<string | null>;

export interface TelegramMessageContentOptions {
  mode: MessageContentMode;
  transformText?: TelegramTextTransformer;
  /** Maximum time allowed for each async transformation. Default `1000`. */
  transformTimeoutMs?: number;
}

/** Framework-neutral privacy contract for Telegram SDK adapters. */
export interface TelegramPrivacyOptions {
  /** Optional Telegram profile fields allowed to leave the bot process. User ids are always retained. */
  contactFields?: readonly TelegramContactField[];
  /** Incoming and observed outgoing content shared with FlowCastle. */
  messageContent?: MessageContentMode | TelegramMessageContentOptions;
}

export interface ResolvedTelegramPrivacyPolicy {
  readonly contactFields: readonly TelegramContactField[];
  readonly messageContent: Readonly<{
    mode: MessageContentMode;
    transformText?: TelegramTextTransformer;
    transformTimeoutMs: number;
  }>;
}

const DEFAULT_TRANSFORM_TIMEOUT_MS = 1000;
const ALL_CONTACT_FIELDS: readonly TelegramContactField[] = [...TELEGRAM_CONTACT_FIELDS];
const CONTACT_KEYS: Readonly<Record<TelegramContactField, string>> = {
  username: 'username',
  firstName: 'first_name',
  lastName: 'last_name',
  languageCode: 'language_code',
  isPremium: 'is_premium',
  addedToAttachmentMenu: 'added_to_attachment_menu',
};

const ENTITY_FIELDS = new Set(['entities', 'caption_entities']);
const STRUCTURED_CONTENT_FIELDS = new Set([
  'animation',
  'audio',
  'contact',
  'dice',
  'document',
  'game',
  'giveaway',
  'giveaway_created',
  'giveaway_completed',
  'giveaway_winners',
  'invoice',
  'location',
  'options',
  'order_info',
  'paid_media',
  'passport_data',
  'photo',
  'poll',
  'poll_answer',
  'proximity_alert_triggered',
  'shipping_address',
  'story',
  'successful_payment',
  'venue',
  'video',
  'video_note',
  'voice',
  'web_app_data',
]);
const MESSAGE_CONTENT_FIELDS = new Set([
  'address',
  'author_signature',
  'bio',
  'caption',
  'caption_entities',
  'city',
  'connected_website',
  'country_code',
  'currency',
  'description',
  'email',
  'entities',
  'explanation',
  'file_name',
  'forward_sender_name',
  'invoice_payload',
  'label',
  'latitude',
  'longitude',
  'horizontal_accuracy',
  'live_period',
  'heading',
  'proximity_alert_radius',
  'name',
  'new_chat_title',
  'paid_media_payload',
  'phone_number',
  'post_code',
  'question',
  'query',
  'quote',
  'shipping_option_id',
  'start_parameter',
  'state',
  'street_line1',
  'street_line2',
  'text',
  'title',
  'total_amount',
  'url',
  'vcard',
]);

const OUTGOING_OPERATIONAL_FIELDS = new Set([
  'action',
  'business_connection_id',
  'callback_query_id',
  'chat_id',
  'direct_messages_topic_id',
  'from_chat_id',
  'inline_message_id',
  'message_id',
  'message_thread_id',
  'pre_checkout_query_id',
  'sender_chat_id',
  'shipping_query_id',
  'until_date',
  'user_id',
]);
const INCOMING_OPERATIONAL_FIELDS = new Set([
  'actor_chat',
  'actor_user',
  'business_connection_id',
  'callback_query',
  'chat',
  'chat_instance',
  'date',
  'edit_date',
  'from',
  'id',
  'is_bot',
  'is_member',
  'message',
  'new_chat_member',
  'old_chat_member',
  'reply_to_message',
  'sender_chat',
  'status',
  'type',
  'user',
  'via_bot',
]);
const INCOMING_OPERATIONAL_ARRAY_FIELDS = new Set(['message_ids']);

type MutableJsonObject = Record<string, JsonValue>;

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function cloneJsonObject(value: unknown): MutableJsonObject | undefined {
  if (!isObject(value)) return undefined;
  return JSON.parse(JSON.stringify(value)) as MutableJsonObject;
}

function normalizeTimeout(value: number | undefined): number {
  if (value === undefined || !Number.isFinite(value) || value < 1) return DEFAULT_TRANSFORM_TIMEOUT_MS;
  return Math.floor(value);
}

/**
 * Resolves the public privacy options. Omitting `privacy` preserves pre-policy
 * SDK behavior; explicitly passing `{}` opts into minimal profile data and
 * routing-only message content.
 */
export function resolveTelegramPrivacyPolicy(
  options?: TelegramPrivacyOptions,
  legacyRedactText = false,
): ResolvedTelegramPrivacyPolicy {
  const content = options?.messageContent;
  const contentOptions: TelegramMessageContentOptions = typeof content === 'string'
    ? { mode: content }
    : content ?? { mode: options === undefined && !legacyRedactText ? 'full' : 'routing' };
  const requestedContactFields = options === undefined
    ? ALL_CONTACT_FIELDS
    : options.contactFields ?? [];
  const requestedSet = new Set(requestedContactFields);

  return {
    contactFields: TELEGRAM_CONTACT_FIELDS.filter((field) => requestedSet.has(field)),
    messageContent: {
      mode: contentOptions.mode,
      ...(contentOptions.transformText === undefined ? {} : { transformText: contentOptions.transformText }),
      transformTimeoutMs: normalizeTimeout(contentOptions.transformTimeoutMs),
    },
  };
}

function updateType(update: MutableJsonObject): string {
  return Object.keys(update).find((key) => key !== 'update_id') ?? 'unknown';
}

function isInlineButton(value: MutableJsonObject): boolean {
  return 'callback_data' in value || 'url' in value || 'web_app' in value || 'login_url' in value;
}

function commandParts(value: string): { command: string; payload?: string } | undefined {
  const match = value.match(/^(\/[A-Za-z0-9_]+(?:@[A-Za-z0-9_]+)?)(?:\s+([\s\S]*))?$/);
  if (match === null) return undefined;
  return { command: match[1], ...(match[2] === undefined ? {} : { payload: match[2] }) };
}

function slimInlineKeyboard(value: MutableJsonObject): void {
  const replyMarkup = value.reply_markup;
  if (!isObject(replyMarkup) || !Array.isArray(replyMarkup.inline_keyboard)) {
    delete value.reply_markup;
    return;
  }
  const rows: JsonValue[][] = [];
  for (const row of replyMarkup.inline_keyboard) {
    if (!Array.isArray(row)) continue;
    const buttons: JsonValue[] = [];
    for (const candidate of row) {
      if (!isObject(candidate)) continue;
      const button: MutableJsonObject = {};
      if (typeof candidate.text === 'string') button.text = candidate.text;
      if (typeof candidate.callback_data === 'string') button.callback_data = candidate.callback_data;
      if (typeof candidate.url === 'string') button.url = candidate.url;
      buttons.push(button);
    }
    rows.push(buttons);
  }
  value.reply_markup = { inline_keyboard: rows };
}

export class TelegramPrivacyFilter {
  public readonly policy: ResolvedTelegramPrivacyPolicy;

  public constructor(
    options?: TelegramPrivacyOptions,
    private readonly onError: (error: unknown) => void = () => undefined,
    legacyRedactText = false,
  ) {
    this.policy = resolveTelegramPrivacyPolicy(options, legacyRedactText);
  }

  public async sanitizeUpdate(update: JsonObject): Promise<JsonObject> {
    const clone = cloneJsonObject(update) ?? {};
    const restricted = this.policy.messageContent.mode === 'full'
      ? clone
      : this.restrictIncomingUpdate(clone);
    await this.sanitizeObject(restricted, [], updateType(restricted));
    return restricted;
  }

  private restrictIncomingUpdate(value: MutableJsonObject): MutableJsonObject {
    const restricted: MutableJsonObject = {};
    if (typeof value.update_id === 'number') restricted.update_id = value.update_id;
    for (const [key, candidate] of Object.entries(value)) {
      if (key === 'update_id') continue;
      if (isObject(candidate)) restricted[key] = this.restrictIncomingObject(candidate as MutableJsonObject);
    }
    return restricted;
  }

  private restrictIncomingObject(value: MutableJsonObject): MutableJsonObject {
    const restricted: MutableJsonObject = {};
    const allowedContactKeys = new Set(this.policy.contactFields.map((field) => CONTACT_KEYS[field]));
    for (const [key, candidate] of Object.entries(value)) {
      const isIdentifier = key.endsWith('_id');
      const isTimestamp = key.endsWith('_date');
      const isContentRoute = this.policy.messageContent.mode === 'routing'
        && (key === 'text' || key === 'data' || key === 'callback_data');
      const keep = INCOMING_OPERATIONAL_FIELDS.has(key)
        || allowedContactKeys.has(key)
        || isIdentifier
        || isTimestamp
        || isContentRoute;
      if (!keep) continue;

      if (Array.isArray(candidate)) {
        if (INCOMING_OPERATIONAL_ARRAY_FIELDS.has(key)) restricted[key] = candidate as JsonValue[];
      } else if (isObject(candidate)) {
        restricted[key] = this.restrictIncomingObject(candidate as MutableJsonObject);
      } else {
        restricted[key] = candidate as JsonValue;
      }
    }
    return restricted;
  }

  public async sanitizeOutgoingPayload(payload: unknown): Promise<JsonObject | undefined> {
    const clone = cloneJsonObject(payload);
    if (clone === undefined) return undefined;
    slimInlineKeyboard(clone);
    const restricted = this.policy.messageContent.mode === 'full'
      ? clone
      : this.restrictOutgoingPayload(clone);
    await this.sanitizeObject(restricted, [], 'outgoing');
    return restricted;
  }

  private restrictOutgoingPayload(value: MutableJsonObject): MutableJsonObject {
    const restricted: MutableJsonObject = {};
    for (const field of OUTGOING_OPERATIONAL_FIELDS) {
      const candidate = value[field];
      if (candidate !== undefined) restricted[field] = candidate;
    }
    if (this.policy.messageContent.mode === 'routing' && value.reply_markup !== undefined) {
      restricted.reply_markup = value.reply_markup;
    }
    return restricted;
  }

  private async sanitizeObject(value: MutableJsonObject, path: readonly string[], eventType: string): Promise<void> {
    this.sanitizeContactFields(value);
    await this.sanitizeDirectContent(value, path, eventType);

    for (const [key, child] of Object.entries(value)) {
      if (Array.isArray(child)) {
        await this.sanitizeArray(child, [...path, key], eventType);
      } else if (isObject(child)) {
        await this.sanitizeObject(child as MutableJsonObject, [...path, key], eventType);
      }
    }
  }

  private async sanitizeArray(value: JsonValue[], path: readonly string[], eventType: string): Promise<void> {
    for (const item of value) {
      if (Array.isArray(item)) await this.sanitizeArray(item, path, eventType);
      else if (isObject(item)) await this.sanitizeObject(item as MutableJsonObject, path, eventType);
    }
  }

  private sanitizeContactFields(value: MutableJsonObject): void {
    const allowed = new Set(this.policy.contactFields);
    for (const field of TELEGRAM_CONTACT_FIELDS) {
      if (!allowed.has(field)) delete value[CONTACT_KEYS[field]];
    }
  }

  private async sanitizeDirectContent(
    value: MutableJsonObject,
    path: readonly string[],
    eventType: string,
  ): Promise<void> {
    const { mode } = this.policy.messageContent;
    if (mode !== 'full') {
      for (const field of STRUCTURED_CONTENT_FIELDS) delete value[field];
      for (const field of MESSAGE_CONTENT_FIELDS) {
        if (field !== 'text') delete value[field];
      }
      if (mode === 'none') {
        delete value.text;
        delete value.data;
        delete value.callback_data;
        delete value.reply_markup;
        return;
      }

      await this.sanitizeRoutingText(value, eventType);
      await this.sanitizeCallbackValue(value, path, eventType);
      return;
    }

    await this.transformFullText(value, path, eventType);
  }

  private async sanitizeRoutingText(value: MutableJsonObject, eventType: string): Promise<void> {
    if (typeof value.text !== 'string') return;
    if (isInlineButton(value)) {
      value.text = '[redacted]';
      return;
    }
    const command = commandParts(value.text);
    if (command === undefined) {
      delete value.text;
      return;
    }
    value.text = command.command;
    for (const field of ENTITY_FIELDS) delete value[field];
    void eventType;
  }

  private async sanitizeCallbackValue(
    value: MutableJsonObject,
    path: readonly string[],
    eventType: string,
  ): Promise<void> {
    if (typeof value.callback_data === 'string') {
      const transformed = await this.transform(value.callback_data, 'callbackData', eventType);
      if (transformed === undefined) delete value.callback_data;
      else value.callback_data = transformed;
    }
    if (path[path.length - 1] === 'callback_query' && typeof value.data === 'string') {
      const transformed = await this.transform(value.data, 'callbackData', eventType);
      if (transformed === undefined) delete value.data;
      else value.data = transformed;
    }
  }

  private async transformFullText(
    value: MutableJsonObject,
    path: readonly string[],
    eventType: string,
  ): Promise<void> {
    if (typeof value.text === 'string') {
      if (isInlineButton(value)) {
        await this.assignTransformed(value, 'text', 'buttonText', eventType);
      } else {
        const command = commandParts(value.text);
        if (command?.payload !== undefined) {
          const payload = await this.transform(command.payload, 'commandArguments', eventType);
          if (payload === undefined || payload.length === 0) value.text = command.command;
          else value.text = `${command.command} ${payload}`;
          for (const field of ENTITY_FIELDS) delete value[field];
        } else {
          await this.assignTransformed(value, 'text', 'messageText', eventType, 'entities');
        }
      }
    }
    await this.assignTransformed(value, 'caption', 'caption', eventType, 'caption_entities');
    await this.assignTransformed(value, 'vcard', 'contactVCard', eventType);
    await this.sanitizeCallbackValue(value, path, eventType);
    if ((path[path.length - 1] === 'inline_query' || path[path.length - 1] === 'chosen_inline_result')) {
      await this.assignTransformed(value, 'query', 'inlineQuery', eventType);
    }
  }

  private async assignTransformed(
    value: MutableJsonObject,
    key: string,
    field: TelegramTextField,
    eventType: string,
    entityKey?: string,
  ): Promise<void> {
    const current = value[key];
    if (typeof current !== 'string') return;
    const transformed = await this.transform(current, field, eventType);
    if (transformed === undefined) delete value[key];
    else value[key] = transformed;
    if (this.policy.messageContent.transformText !== undefined && entityKey !== undefined) delete value[entityKey];
  }

  private async transform(
    value: string,
    field: TelegramTextField,
    eventType: string,
  ): Promise<string | undefined> {
    const transformer = this.policy.messageContent.transformText;
    if (transformer === undefined) return value;

    let timeout: ReturnType<typeof setTimeout> | undefined;
    try {
      const timeoutPromise = new Promise<never>((_resolve, reject) => {
        timeout = setTimeout(
          () => reject(new Error(`transformText timed out after ${this.policy.messageContent.transformTimeoutMs}ms`)),
          this.policy.messageContent.transformTimeoutMs,
        );
      });
      const transformed = await Promise.race([
        Promise.resolve(transformer({ value, field, updateType: eventType })),
        timeoutPromise,
      ]);
      if (transformed === null) return undefined;
      if (typeof transformed !== 'string') throw new TypeError('transformText must return a string or null');
      return transformed;
    } catch (error) {
      const detail = error instanceof Error ? error.message : String(error);
      this.report(new Error(`FlowCastle: transformText failed for ${field}; content dropped: ${detail}`));
      return undefined;
    } finally {
      if (timeout !== undefined) clearTimeout(timeout);
    }
  }

  private report(error: unknown): void {
    try {
      this.onError(error);
    } catch {
      // Privacy failures must not escape through observability callbacks.
    }
  }
}

File: packages/sdk-python/src/flowcastle/privacy.py
"""Fail-closed Telegram payload privacy filtering."""

from __future__ import annotations

import asyncio
import copy
import inspect
import re
from dataclasses import dataclass
from typing import Awaitable, Callable, Literal, Mapping, TypedDict, cast

from .types import JsonObject

ContentMode = Literal['full', 'routing', 'none']
CONTACT_FIELDS = ('username', 'first_name', 'last_name', 'language_code', 'is_premium', 'added_to_attachment_menu')
CONTENT_FIELDS = {
    'address', 'caption', 'caption_entities', 'contact', 'document', 'email', 'entities',
    'file_name', 'location', 'phone_number', 'photo', 'poll', 'text', 'venue', 'video',
    'voice', 'vcard', 'query', 'data', 'callback_data',
}
OPERATIONAL_FIELDS = {
    'actor_chat', 'business_connection_id', 'callback_query', 'chat', 'chat_instance', 'date',
    'edit_date', 'from', 'id', 'is_bot', 'is_member', 'message', 'new_chat_member',
    'old_chat_member', 'reply_to_message', 'sender_chat', 'status', 'type', 'user', 'via_bot',
}
OUTGOING_OPERATIONAL = {
    'action', 'business_connection_id', 'callback_query_id', 'chat_id', 'from_chat_id',
    'inline_message_id', 'message_id', 'message_thread_id', 'pre_checkout_query_id',
    'sender_chat_id', 'shipping_query_id', 'until_date', 'user_id',
}


class TextTransformContext(TypedDict):
    value: str
    field: str
    update_type: str


TextTransformer = Callable[[TextTransformContext], str | None | Awaitable[str | None]]


@dataclass(frozen=True)
class PrivacyOptions:
    contact_fields: tuple[str, ...] | None = None
    message_content: ContentMode = 'routing'
    transform_text: TextTransformer | None = None
    transform_timeout_ms: int = 1000


class PrivacyFilter:
    def __init__(self, options: PrivacyOptions | Mapping[str, object] | None, on_error: Callable[[Exception], None]) -> None:
        self._options = self._normalize(options)
        self._on_error = on_error
        # Compatibility: no policy requested retains previous complete payload behavior.
        self._mode: ContentMode = 'full' if options is None else self._options.message_content
        self._contacts = set(CONTACT_FIELDS if options is None else self._options.contact_fields or ())

    def _normalize(self, options: PrivacyOptions | Mapping[str, object] | None) -> PrivacyOptions:
        if options is None:
            return PrivacyOptions(tuple(CONTACT_FIELDS), 'full')
        if isinstance(options, PrivacyOptions):
            return options
        raw_contacts = options.get('contact_fields')
        contacts = tuple(value for value in raw_contacts if isinstance(value, str)) if isinstance(raw_contacts, (list, tuple)) else ()
        raw_mode = options.get('message_content', 'routing')
        mode: ContentMode = raw_mode if raw_mode in {'full', 'routing', 'none'} else 'routing'
        transformer = options.get('transform_text')
        timeout = options.get('transform_timeout_ms', 1000)
        return PrivacyOptions(
            contacts,
            mode,
            cast(TextTransformer | None, transformer if callable(transformer) else None),
            timeout if isinstance(timeout, int) else 1000,
        )

    async def sanitize_update(self, update: JsonObject) -> JsonObject:
        clone = copy.deepcopy(update)
        if self._mode != 'full':
            clone = self._restrict_incoming(clone)
        await self._sanitize_recursive(clone, self._update_type(clone), ())
        return clone

    async def sanitize_outgoing(self, payload: JsonObject) -> JsonObject:
        clone = copy.deepcopy(payload)
        if self._mode != 'full':
            clone = {key: value for key, value in clone.items() if key in OUTGOING_OPERATIONAL}
            if self._mode == 'routing' and isinstance(payload.get('reply_markup'), dict):
                clone['reply_markup'] = self._slim_keyboard(payload['reply_markup'])
        await self._sanitize_recursive(clone, 'outgoing', ())
        return clone

    def _restrict_incoming(self, update: JsonObject) -> JsonObject:
        result: JsonObject = {}
        if isinstance(update.get('update_id'), (str, int)):
            result['update_id'] = update['update_id']
        for key, value in update.items():
            if key != 'update_id' and isinstance(value, dict):
                result[key] = self._restrict_object(value)
        return result

    def _restrict_object(self, value: JsonObject) -> JsonObject:
        result: JsonObject = {}
        for key, candidate in value.items():
            route_content = self._mode == 'routing' and key in {'text', 'data', 'callback_data'}
            allowed = key in OPERATIONAL_FIELDS or key in self._contacts or key.endswith('_id') or key.endswith('_date') or route_content
            if not allowed:
                continue
            if isinstance(candidate, dict):
                result[key] = self._restrict_object(candidate)
            elif isinstance(candidate, list):
                if key == 'message_ids':
                    result[key] = candidate
            else:
                result[key] = candidate
        return result

    async def _sanitize_recursive(self, value: JsonObject, update_type: str, path: tuple[str, ...]) -> None:
        for contact in CONTACT_FIELDS:
            if contact not in self._contacts:
                value.pop(contact, None)
        if self._mode == 'none':
            for field in CONTENT_FIELDS | {'reply_markup'}:
                value.pop(field, None)
        elif self._mode == 'routing':
            await self._routing_content(value, update_type, path)
        else:
            await self._full_content(value, update_type, path)
        for key, child in list(value.items()):
            if isinstance(child, dict):
                await self._sanitize_recursive(child, update_type, path + (key,))
            elif isinstance(child, list):
                for entry in child:
                    if isinstance(entry, dict):
                        await self._sanitize_recursive(entry, update_type, path + (key,))

    async def _routing_content(self, value: JsonObject, update_type: str, path: tuple[str, ...]) -> None:
        for field in CONTENT_FIELDS - {'text', 'data', 'callback_data'}:
            value.pop(field, None)
        text = value.get('text')
        if isinstance(text, str):
            command = re.match(r'^(\/[A-Za-z0-9_]+(?:@[A-Za-z0-9_]+)?)(?:\s+.*)?$', text)
            if command is None:
                value.pop('text', None)
            else:
                value['text'] = command.group(1)
                value.pop('entities', None)
        await self._transform_callback(value, update_type, path)

    async def _full_content(self, value: JsonObject, update_type: str, path: tuple[str, ...]) -> None:
        for key, field in (('text', 'message_text'), ('caption', 'caption'), ('vcard', 'contact_vcard')):
            await self._transform_field(value, key, field, update_type)
        await self._transform_callback(value, update_type, path)

    async def _transform_callback(self, value: JsonObject, update_type: str, path: tuple[str, ...]) -> None:
        for key in ('callback_data', 'data'):
            if key == 'data' and (not path or path[-1] != 'callback_query'):
                continue
            await self._transform_field(value, key, 'callback_data', update_type)

    async def _transform_field(self, value: JsonObject, key: str, field: str, update_type: str) -> None:
        original = value.get(key)
        if not isinstance(original, str) or self._options is None or self._options.transform_text is None:
            return
        try:
            transformed = self._options.transform_text({'value': original, 'field': field, 'update_type': update_type})
            if inspect.isawaitable(transformed):
                transformed = await asyncio.wait_for(transformed, self._timeout_seconds())
            if transformed is None:
                value.pop(key, None)
            elif isinstance(transformed, str):
                value[key] = transformed
            else:
                raise TypeError('transform_text must return str or None')
        except Exception as error:
            value.pop(key, None)
            self._report(error)

    def _timeout_seconds(self) -> float:
        assert self._options is not None
        return max(1, self._options.transform_timeout_ms) / 1000

    def _update_type(self, update: JsonObject) -> str:
        return next((key for key in update if key != 'update_id'), 'unknown')

    def _slim_keyboard(self, value: object) -> JsonObject:
        if not isinstance(value, dict) or not isinstance(value.get('inline_keyboard'), list):
            return {}
        rows: list[list[JsonObject]] = []
        for row in value['inline_keyboard']:
            if not isinstance(row, list):
                continue
            buttons = [{key: item[key] for key in ('text', 'callback_data', 'url') if isinstance(item, dict) and key in item} for item in row]
            rows.append(buttons)
        return {'inline_keyboard': rows}

    def _report(self, error: Exception) -> None:
        try:
            self._on_error(error)
        except Exception:
            pass
